## Lab 1 - Basic NumPy Operations

This notebook demonstrates basic NumPy operations,
including arrays, slicing, stacking, broadcasting,
and linear algebra.

## Part A

### Import Numpy :
Below We will Import Numpy

In [2]:
import numpy as np

### Question 1 :
Create an array that starts from integer 1, ends at 20, incremented by 3.

In [3]:
array1 = np.arange(1, 21, 3)

print(array1)

[ 1  4  7 10 13 16 19]


### Question 2
Create a new array of shape 3 with random numbers between 0 and 1.

In [4]:
random_array = np.random.rand(3)

print(random_array)

[0.4938806  0.87719904 0.90398266]


### Question 3:
Create a 2 D array [[10,20,45], [30,12,16], [42,17,56]] 

In [5]:
array2 = np.array([
    [10, 20, 45],
    [30, 12, 16],
    [42, 17, 56]
])

print(array2)

[[10 20 45]
 [30 12 16]
 [42 17 56]]


Get First two rows

In [6]:
first_two_rows = array2[:2]

print(first_two_rows)

[[10 20 45]
 [30 12 16]]


Get Last two rows

In [7]:
last_two_rows = array2[-2:]

print(last_two_rows)

[[30 12 16]
 [42 17 56]]


### Question

In [8]:
a = np.array([
    [1, 2],
    [3, 4]
])

b = np.array([
    [5, 6],
    [7, 8]
])

print(a)
print(b)

[[1 2]
 [3 4]]
[[5 6]
 [7 8]]


### Vertical Stacking : 
Putting a over b

In [9]:
vertical = np.vstack((a, b))

print(vertical)

[[1 2]
 [3 4]
 [5 6]
 [7 8]]


### Horizontal stacking :
Now put a beside b

In [10]:
horizontal = np.hstack((a, b))

print(horizontal)

[[1 2 5 6]
 [3 4 7 8]]


### Vertical Splitting
Breaking arrays apart vertically

In [11]:
split_array = np.array([
    [1, 2],
    [3, 4],
    [5, 6],
    [7, 8]
])

result = np.vsplit(split_array, 2)

print(result)

[array([[1, 2],
       [3, 4]]), array([[5, 6],
       [7, 8]])]


### Horizontal Splitting
Breaking arrays apart Horizontally

In [12]:
horizontal_split_array = np.array([
    [1, 2, 3, 4],
    [5, 6, 7, 8]
])

result2 = np.hsplit(horizontal_split_array, 2)

print(result2)

[array([[1, 2],
       [5, 6]]), array([[3, 4],
       [7, 8]])]


### Question 

Lets create arrays X and Y

In [13]:
X = np.array([
    [5, 7, 2],
    [4, 5, 6],
    [7, 4, 2]
])

Y = np.array([
    [4, 2],
    [6, 2],
    [4, 2]
])

To see if the multiplication of X and Y is possible

In [14]:
result = X @ Y

print(result)

[[70 28]
 [70 30]
 [60 26]]


### Matrix Multiplication Possible

Matrix X has shape (3,3) and matrix Y has shape (3,2).Therefore, Matrix multiplication is possible because the inner dimensions are equal (3 and 3), and the resulting matrix has shape (3,2).



In [15]:
print("Shape of X:", X.shape)
print("Shape of Y:", Y.shape)

Shape of X: (3, 3)
Shape of Y: (3, 2)


### Matrix Multiplication not Possible
For demonstration, Y_wrong has shape (2,2). Multiplying X (3,3) by Y_2 (2,2) is not possible because the inner dimensions 3 and 2 do not match.

In [16]:
#Let's create one more Y_2 array with shape (2, 2) and try multiply it with X
Y_2 = np.array([
    [4, 2],
    [6, 2]
])
print("Shape of X:", X.shape)
print("Shape of Y_2:", Y_2.shape)



Shape of X: (3, 3)
Shape of Y_2: (2, 2)


### Question
Create two arrays, x = ([2, -1, -8]) y = ([3, 1, -2]), Find the Shape, Number of dimensions of vector x.
     
Reshape the vector x to a matrix of size (3,1) and determine the number of dimensions after reshaping y to a matrix of (3,1).

In [17]:
x = np.array([2, -1, -8])
y = np.array([3, 1, -2])

To find shape and number of dimentions

In [18]:
print("Shape of x:", x.shape)
print("Number of dimensions of x:", x.ndim)

Shape of x: (3,)
Number of dimensions of x: 1


Reshape x to a matrix of size (3,1).

In [19]:
x_matrix = x.reshape(3, 1)

print(x_matrix)

[[ 2]
 [-1]
 [-8]]


To check updated shape and number of vectors

In [20]:
print("Shape:", x_matrix.shape)
print("Number of dimensions:", x_matrix.ndim)

Shape: (3, 1)
Number of dimensions: 2


determining the number of dimensions after reshaping y to (3,1).

In [21]:
y_matrix = y.reshape(3, 1)

print(y_matrix)
print("Number of dimensions:", y_matrix.ndim)

[[ 3]
 [ 1]
 [-2]]
Number of dimensions: 2


### Question
How does broadcasting work? Demonstrate the subtraction, multiplication by considering a 3 x 3 matrix.

**Broadcasting** is  automatically applying a number to every element of matrices. **For Example** Lets create a 3*3 matrix and consider our number is 2

In [22]:
matrix = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
])

number = 2

#### Broadcasting Subtraction 
It will subtract number(2) from matrix

In [23]:
subtraction = matrix - number

print(subtraction)

[[-1  0  1]
 [ 2  3  4]
 [ 5  6  7]]


#### Broadcasting multiplication 
It will multiply number(2) from matrix

In [24]:
multiplication = matrix * number

print(multiplication)

[[ 2  4  6]
 [ 8 10 12]
 [14 16 18]]


# Overview

This notebook demonstrates basic NumPy operations using Python.

- **Array Creation:** NumPy can quickly create numerical arrays.
- **Indexing and Slicing:** Specific rows and elements can be selected from arrays.
- **Stacking and Splitting:** Arrays can be combined or divided using NumPy functions.
- **Matrix Multiplication:** NumPy can perform matrix multiplication when the matrix dimensions are compatible.
- **Reshaping:** Arrays can be changed into different shapes without changing their values.
- **Broadcasting:** NumPy can perform operations between arrays of compatible shapes automatically.
- **Linear Algebra:** NumPy provides tools for solving systems of linear equations.

## Part B

### System 1
Solve the following two system of linear equations using matrices (find the values of x1 and x2) and 
      determine the number of solutions using the functions.

         1.2x1+3x2-4x3=6,x1-4x2=8
	
         2. 3y1-4y2+5y3=10,-y1+2y2-4y3=8

#### Solution
The rank of the coefficient matrix and the augmented matrix is used to determine the number of solutions.We need to convert equations into matrices( A1X = B1). Which gives us A1 =
[[ 2,  3, -4],
 [ 1, -4,  0]] and
  B1 =
[6, 8] and
X = [x1, x2, x3]

In [25]:
A1 = np.array([
    [2, 3, -4],
    [1, -4, 0]
])

B1 = np.array([6, 8])

print("Coefficient Matrix A:")
print(A1)

print("\nConstant Matrix B:")
print(B1)

Coefficient Matrix A:
[[ 2  3 -4]
 [ 1 -4  0]]

Constant Matrix B:
[6 8]


To find the solution

In [26]:

# STEP 1: Define the Coefficient Matrix (A) and Vector (B) 
# Setting x3 = 0 to get 2 equations for 2 variables (x1, x2)
A1 = np.array([[2, 3],
               [1, -4]])

B1 = np.array([6, 8])

#  STEP 2: Check Number of Solutions using Rank 
# Build full matrix including x3 and constant term
A1_full = np.array([[2, 3, -4],
                    [1, -4, 0]])

A1_aug = np.array([[2, 3, -4, 6],
                   [1, -4, 0, 8]])

rank_A1 = np.linalg.matrix_rank(A1_full)
rank_aug1 = np.linalg.matrix_rank(A1_aug)

print("Rank of A:", rank_A1)
print("Rank of Augmented:", rank_aug1)

if rank_A1 == rank_aug1:
    print("Number of solutions: Infinitely many solutions (since rank < 3 variables)")

# --- STEP 3: Solve for x1 and x2 ---
# Solve A1 * X = B1 assuming x3 = 0
X1 = np.linalg.solve(A1, B1)

print("Value of x1:", X1[0])
print("Value of x2:", X1[1])

Rank of A: 2
Rank of Augmented: 2
Number of solutions: Infinitely many solutions (since rank < 3 variables)
Value of x1: 4.363636363636363
Value of x2: -0.9090909090909092


### Result for System 1

Number of solutions: Infinitely many solutions   
Values (when $x_3 = 0$): $x_1 \approx 4.36$, $x_2 \approx -0.91$

### System 2

The system of equations is:

3y1 - 4y2 + 5y3 = 10

-y1 + 2y2 - 4y3 = 8

Using similar coefficient technique as used in system 1

In [28]:
A2 = np.array([
    [3, -4, 5],
    [-1, 2, -4]
])

B2 = np.array([10, 8])

print("Coefficient Matrix A:")
print(A2)

print("\nConstant Matrix B:")
print(B2)

Coefficient Matrix A:
[[ 3 -4  5]
 [-1  2 -4]]

Constant Matrix B:
[10  8]


To find the solution

In [ ]:
# STEP 1: Define the Coefficient Matrix (A) and Vector (B) 
# Setting y3 = 0 to get 2 equations for 2 variables (y1, y2)
A2 = np.array([[ 3, -4],
               [-1,  2]])

B2 = np.array([10, 8])

# --- STEP 2: Check Number of Solutions using Rank ---
A2_full = np.array([[ 3, -4, 5],
                    [-1,  2, -4]])

A2_aug = np.array([[ 3, -4, 5, 10],
                   [-1,  2, -4, 8]])

rank_A2 = np.linalg.matrix_rank(A2_full)
rank_aug2 = np.linalg.matrix_rank(A2_aug)

print("Rank of A:", rank_A2)
print("Rank of Augmented:", rank_aug2)

if rank_A2 == rank_aug2:
    print("Number of solutions: Infinitely many solutions (since rank < 3 variables)")

# --- STEP 3: Solve for y1 and y2 ---
# Solve A2 * Y = B2 assuming y3 = 0
Y2 = np.linalg.solve(A2, B2)

print("Value of y1:", Y2[0])
print("Value of y2:", Y2[1])

Rank of A: 2
Rank of Augmented: 2
Number of solutions: Infinitely many solutions (since rank < 3 variables)
Value of y1: 26.0
Value of y2: 17.0


### Result for System 2

Number of solutions: Infinitely many solutions   Values (when $y_3 = 0$): $y_1 = 26$, $y_2 = 17$